### ISSUE TITLE
bn256: when to use constants.p (if at all)

### ISSUE THREAD

**Issue by tharvik** [dedis/kyber#412]
Disclaimer: I don't know much about crypto, I might be completely wrong.

Since dedis/cothority#2175, I'm wondering why `p` (the prime which was confused with the order) is needed at all. In golang, it is used solely in `hashToPoint` (which is used by `Hash`), again to do a modulo, which looks wrong, as it would cover a smaller field in my understanding (which is prime also, so the properties are similar ie hard to detect). Replacing the references to it with `Order` doesn't change the behavior of the point, solely the hashes. So, do we need `p` at all?

Now, there is a similar issue in typescript, excepted that `p` is wildly more used; but there, blindly replacing `p` with `order` doesn't works, so I guess I'm missing something.

**Comment by ineiti** [comment]
I really need to do my `crypto for engineers` with links to lots of resources ;)

TLDR: yes, we _do_ need to calculate modulo `p`.  

Proof by implementation: 
- all operations end with a `Mod` here: https://github.com/dedis/kyber/blob/master/group/mod/int.go
- Reference implementation of ed25519: https://ed25519.cr.yp.to/software.html

Proof by links:
- Nice introduction that explains how to calculate elliptic curves: https://andrea.corbellini.name/2015/05/23/elliptic-curve-cryptography-finite-fields-and-discrete-logarithms/
- Original paper: https://ed25519.cr.yp.to/ed25519-20110926.pdf

**Comment by tharvik** [comment]
> I really need to do my `crypto for engineers` with links to lots of resources ;)

I'll be the first to read it :)

> * field: If I understand correctly, elliptic curves use 2 finite fields defined over the modulo operation: the scalars with size _prime order_, and the points with size _p_.
> * p (2**255-19): used to restrict the results of the calculation in the elliptic curve to values that can be handled more quickly by computers.
> * prime order (< p): linked to the number of points you can generate using the base-point.

Haa, I think I got it now, even if it's an engineer intution; thanks for the clarification.
